lib/Semantic_check.ml

```diff
@@ -1,2163 +0,0 @@
-   have a block level that is determined by what has been assigned to them
-   rather than by where they were declared. I'm not sure that behaviour makes
-   sense unless we use static analysis as well to make sure these assignments
-   actually get evaluated in that phase. *)
-let update_originblock name ob =
-  match Symbol_table.look vm name with
-  | Some (old_ob, ut) ->
-      let new_ob = lub_originblock [ob; old_ob] in
-      Symbol_table.unsafe_replace vm name (new_ob, ut)
-  | _ -> fatal_error ()
-
-(* The actual semantic checks for all AST nodes! *)
-let rec semantic_check_program p =
-  match p
-  with
-  | { functionblock= bf
-    ; datablock= bd
-    ; transformeddatablock= btd
-    ; parametersblock= bp
-    ; transformedparametersblock= btp
-    ; modelblock= bm
-    ; generatedquantitiesblock= bgq }
-  ->
-    (* NB: We always want to make sure we start with an empty symbol table, in
-       case we are processing multiple files in one run. *)
-    let _ = unsafe_clear_symbol_table vm in
-    let _ = context_flags.current_block <- Functions in
-    let ubf =
-      Core_kernel.Option.map ~f:(List.map semantic_check_statement) bf
-    in
-    let _ =
-      if
-        Symbol_table.check_some_id_is_unassigned vm
-        && !check_that_all_functions_have_definition
-      then
-        semantic_error
-          ~loc:
-            (snd
-               (typed_statement_unroll
-                  (List.hd ((function Some x -> x | _ -> fatal_error ()) ubf))))
-              .stmt_typed_meta_loc
-          "Some function is declared without specifying a definition."
-      (* TODO: insert better location in the error above *)
-    in
-    let _ = context_flags.current_block <- Data in
-    let ubd =
-      Core_kernel.Option.map ~f:(List.map semantic_check_statement) bd
-    in
-    let _ = context_flags.current_block <- TData in
-    let ubtd =
-      Core_kernel.Option.map ~f:(List.map semantic_check_statement) btd
-    in
-    let _ = context_flags.current_block <- Param in
-    let ubp =
-      Core_kernel.Option.map ~f:(List.map semantic_check_statement) bp
-    in
-    let _ = context_flags.current_block <- TParam in
-    let ubtp =
-      Core_kernel.Option.map ~f:(List.map semantic_check_statement) btp
-    in
-    let _ = context_flags.current_block <- Model in
-    let _ = Symbol_table.begin_scope vm in
-    let ubm =
-      Core_kernel.Option.map ~f:(List.map semantic_check_statement) bm
-    in
-    let _ = Symbol_table.end_scope vm in
-    let _ = context_flags.current_block <- GQuant in
-    let ubgq =
-      Core_kernel.Option.map ~f:(List.map semantic_check_statement) bgq
-    in
-    { functionblock= ubf
-    ; datablock= ubd
-    ; transformeddatablock= ubtd
-    ; parametersblock= ubp
-    ; transformedparametersblock= ubtp
-    ; modelblock= ubm
-    ; generatedquantitiesblock= ubgq }
-
-(* This could also be dealt with during lexing. That would probably be more efficient. *)
```

Comment:
should we do this instead?

Reply:
I don't think so actually. I take back that comment. I don't see how to throw good errors as easily that way. If you feel like rewriting it the other way, go ahead though!